File: src/main/java/com/github/alikemalocalan/instastorysaver/SerializableCookieJar.java
package com.github.alikemalocalan.instastorysaver;

import okhttp3.Cookie;
import okhttp3.CookieJar;
import okhttp3.HttpUrl;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class SerializableCookieJar implements CookieJar, Serializable {

    private static final long serialVersionUID = -837498359387593793L;

    Map<String, List<SerializableCookie>> map = new HashMap<>();

    @Override
    public List<Cookie> loadForRequest(HttpUrl arg0) {
        return map.getOrDefault(arg0.host(), new ArrayList<>()).stream()
                .map(c -> c.cookie)
                .collect(Collectors.toList());
    }

    @Override
    public void saveFromResponse(HttpUrl arg0, List<Cookie> arg1) {
        List<SerializableCookie> list =
                arg1.stream().map(SerializableCookie::new).collect(Collectors.toList());
        if (map.containsKey(arg0.host())) {
            map.get(arg0.host()).addAll(list);
        } else map.put(arg0.host(), list);
    }


    public static class SerializableCookie implements Serializable {

        public SerializableCookie(Cookie cookie) {
            this.cookie = cookie;
        }

        private static final long serialVersionUID = -8594045714036645534L;

        private transient Cookie cookie;

        private void writeObject(ObjectOutputStream out) throws IOException {
            Utils.writeObject(cookie, out);
        }

        private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException {
            cookie = Utils.readObject(in);
        }

    }

}

File: src/main/scala/com/github/alikemalocalan/instastorysaver/model/SerializableCookieJar.scala
package com.github.alikemalocalan.instastorysaver.model

import com.github.alikemalocalan.instastorysaver.Utils
import okhttp3.{Cookie, CookieJar, HttpUrl}

import java.io.{ObjectInputStream, ObjectOutputStream}
import java.util
import scala.collection.mutable
import scala.collection.mutable.ListBuffer
import scala.jdk.CollectionConverters._

class SerializableCookieJar extends CookieJar with Serializable {

  val map: mutable.Map[String, ListBuffer[SerializableCookie]] = mutable.Map[String, ListBuffer[SerializableCookie]]()

  override def loadForRequest(httpUrl: HttpUrl): util.List[Cookie] = {
    map.getOrElse(httpUrl.host(), ListBuffer[SerializableCookie]().empty)
      .map(c => c.cookie)
      .asJava
  }

  override def saveFromResponse(httpUrl: HttpUrl, list: java.util.List[Cookie]): Unit = {
    val cookies = list.asScala.map(SerializableCookie).to(ListBuffer)

    if (map.contains(httpUrl.host())) {
      map(httpUrl.host()).addAll(cookies)
    } else map.put(httpUrl.host(), cookies)
  }

  case class SerializableCookie(var cookie: Cookie) extends Serializable {

    private def writeObject(out: ObjectOutputStream): Unit = {
      Utils.writeObject(cookie, out)
    }

    private def readObject(in: ObjectInputStream): Unit = {
      cookie = Utils.readObject(in)
    }
  }

}
